metrics/sink.go

```diff
@@ -15,143 +15,186 @@ var (
 	_ Sink = &DummySink{}
 )
 
+// Sink collects Sample's values.
 type Sink interface {
```

Comment:
@codebien, I'll start a thread here to discuss [your proposal](https://github.com/grafana/k6/pull/2898#issuecomment-1484811792) for a more generic `Query` and `QueryArguments`, since I can't reply to it without a new top-level PR comment.

I don't particularly like the current `Sink` APIs and I am open to improvement suggestions. However, I am afraid that what you are proposing will result in even more verbose code :disappointed: Particularly having to deal with error handling everywhere instead of type assertions seems like a worse trade-off, e.g. if you `Query()` for things that a `Sink` doesn't have. Or maybe I am misunderstanding something?

Reply:
At one point I remember as discussing basically dropping the Sink API?

The current api stats call, the summary and the thresholds are basically three "outputs" (for a very generaous definition of outputs) that put stuff in Sink and then use it.

There really isn't a good reason why they should use the same abstraction (or to use different ones if they want to). And even less reason for that abstraction to be in the `metric` package and being part of the metric. 

It might still be beneficial that they have a shared store that the metric samples get "sinked" to. But IMO having that be less generic solution will help a lot.

Arguably the current problem is that the 4 different types of metric sinks are combined under one interface and we try to keep them like that. While if the Store just stores Trends, Coutners, etc. as separate things we might completely circumvent this problem

Reply:
Interesting ideas... :thinking: First, a small correction:
> The current api stats call, the summary and the thresholds are basically three "outputs" (for a very generaous definition of outputs) that put stuff in Sink and then use it.

They are in fact a single output. And, after some of my recent refactoring, the current `MetricsIngester` for all three pretends to be an actual `output.Output`: https://github.com/grafana/k6/blob/124fc7919049803cbf404b345f3b67fed0a57ad8/cmd/run.go#L128-L132

> There really isn't a good reason why they should use the same abstraction (or to use different ones if they want to). 

I am not sure, it seems to me like the [simple `Sink` interface](https://github.com/grafana/k6/blob/83963aec4aaa39cf5c26d0f24f42b64c16ab5ca9/metrics/sink.go#L18-L25) from this PR is should simplify Adding values, since that part is universal. Querying the sinks is the different part, which I am not sure how can be made simpler than this :thinking: Can you elaborate?

> And even less reason for that abstraction to be in the metric package and being part of the metric.

This I fully agree with. A follow-up PR after this one should split away the `metrics.Metric` along the lines of the `TODO`: https://github.com/grafana/k6/blob/124fc7919049803cbf404b345f3b67fed0a57ad8/metrics/metric.go#L11-L26

Everything after it should be moved out of the `Metric` type and into various data structures in the `MetricsEngine`.